src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/UseSpanBasedStringConcat.cs

```diff
@@ -0,0 +1,253 @@
+            var builder = ImmutableArray.CreateBuilder<IOperation>();
+            GoLeft(root);
+
+            while (stack.Count != 0)
+            {
+                var current = stack.Pop();
+
+                if (current.LeftOperand is not IBinaryOperation leftBinary || leftBinary.OperatorKind != root.OperatorKind)
+                {
+                    builder.Add(current.LeftOperand);
+                }
+
+                if (current.RightOperand is not IBinaryOperation rightBinary || rightBinary.OperatorKind != root.OperatorKind)
+                {
+                    builder.Add(current.RightOperand);
+                }
+                else
+                {
+                    GoLeft(rightBinary);
+                }
+            }
+
+            return builder.ToImmutable();
+
+            void GoLeft(IBinaryOperation operation)
+            {
+                IBinaryOperation? current = operation;
+                while (current is not null && current.OperatorKind == root.OperatorKind)
+                {
+                    stack.Push(current);
+                    current = current.LeftOperand as IBinaryOperation;
+                }
+            }
+        }
+
+        // Use readonly struct instead of record type to save on allocations, since it's not passed by-value.
+        // We aren't comparing these.
+#pragma warning disable CA1815 // Override equals and operator equals on value types
+        internal readonly struct RequiredSymbols
+#pragma warning restore CA1815 // Override equals and operator equals on value types
+        {
+            private RequiredSymbols(
+                INamedTypeSymbol stringType, INamedTypeSymbol roscharType,
+                IMethodSymbol substring1, IMethodSymbol substring2,
+                IMethodSymbol asSpan1, IMethodSymbol asSpan2)
+            {
+                StringType = stringType;
+                RoscharType = roscharType;
+                Substring1 = substring1;
+                Substring2 = substring2;
+                AsSpan1 = asSpan1;
+                AsSpan2 = asSpan2;
+
+                RoslynDebug.Assert(
+                    StringType is not null && RoscharType is not null &&
+                    Substring1 is not null && Substring2 is not null &&
+                    AsSpan1 is not null && AsSpan2 is not null);
+            }
+
+            public static bool TryGetSymbols(Compilation compilation, out RequiredSymbols symbols)
+            {
+                var stringType = compilation.GetSpecialType(SpecialType.System_String);
+                var roscharType = compilation.GetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemReadOnlySpan1)
+                    ?.Construct(compilation.GetSpecialType(SpecialType.System_Char));
+                var memoryExtensionsType = compilation.GetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemMemoryExtensions);
+
+                if (roscharType is null || memoryExtensionsType is null)
```

Comment:
I'd be more pessimistic and check against `stringType` for null. It's unlikely to be null, but I think it may be.

Reply:
Fixed.